Modernize the test driver

timvandermeij opened the pull request:
_The commit messages contain more information about the individual changes. The `?w=1` flag is most likely necessary for reviewing most commits._

Initially I only tried to enable the `no-var` linting rule since our testing code is one of the final places where we exclude that rule, but I quickly found that the driver code could use further modernization. Hence this commit series, where each commit contains a separate refactoring step. Note that there is most likely more opportunity for improvement here, for instance using template strings more often and getting rid of XHR, but I deliberately left that for a follow-up since this series is already quite a big.

This PR makes the driver code 103 lines shorter and, in my opinion at least, easier to understand and work with.

timvandermeij committed:
Convert the `Driver` class to ES6 syntax in `test/driver.js`

timvandermeij committed:
Use arrow functions instead of `self` variables in `test/driver.js`

timvandermeij commented:
/botio test

Snuffleupagus commented on test/driver.js:
```diff
@@ -143,89 +144,79 @@ async function resolveImages(node, silentErrors = false) {
+  textContent,
+  enhanceTextSelection
+) {
   const styles = {
     common: {
       file: "./text_layer_test.css",
       promise: null,
     },
   };
```
Won't this re-factoring mean that you're now re-initializing the style `promise` for *every* single `rasterizeTextLayer` call, rather than caching it as intended here!?

This probably applies elsewhere below in the file as well.

Snuffleupagus replied:
How about e.g. adding a class with static methods, e.g. something like this since that'd allow you to handle the style-caching internally instead:
```javascript
class Rasterize {
  static #textStyles = {...};

  static #annotationStyles = {...};

  static #xfaStyles = {...};

  static async textLayer(...) {}

  static async annotationLayer(...) {}

  static async xfaLayer(...) {}
}
```

timvandermeij replied:
Ah, you're right. Somehow I managed to completely overlook this. I'll revisit the patch series today.

timvandermeij replied:
Fixed in the new commit series.

Snuffleupagus commented on test/driver.js:
```diff
@@ -428,569 +376,550 @@ var Driver = (function DriverClosure() {
-      const queryString = window.location.search.substring(1);
-      return Object.fromEntries(new URLSearchParams(queryString).entries());
-    },
+  _getQueryStringParameters() {
```
Should we perhaps export https://github.com/mozilla/pdf.js/blob/dc455c836e80799fccd8f1741eb8f62e9ec01c37/web/ui_utils.js#L183-L194 in the viewer components and re-use that in this file rather than essentially re-implementing that here?

timvandermeij replied:
For now I skipped that because the series is already quite big and I intended to only touch the driver code here. We can do that in a follow-up though because I don't see other limitations for us to do that.

Snuffleupagus reviewed (changes requested):
It's really nice that you're doing this, but before I review it in detail there's one thing I fear is incorrect; please see the inline comment.

Snuffleupagus commented:
> and getting rid of XHR,

*Just a word of warning:* I actually tried that recently, but abandoned it for the time being, since I ran into some trouble; please see https://github.com/mozilla/pdf.js/pull/14287#issuecomment-974200678

timvandermeij committed:
Inline `loadStyles` calls in the rasterization classes in `test/driver.js`

The wrapper functions in this case only really added indirection, so
this commit simplifies the code a bit.

timvandermeij committed:
Use async/await in the rasterization classes in `test/driver.js`

This is achieved by letting the `writeSVG` function return a promise so
we don't need callback passing anymore.

timvandermeij committed:
Enable the `no-var` linting rule in `test/driver.js`

This is done automatically with the `gulp lint --fix` command with the
only exception of the `annotationLayerContext` variable.

timvandermeij committed:
Move the rasterization logic into one single class

This refactoring ensures that we can get rid of the closures and
encapsulate the logic in a nicer way with e.g., getters for the style
promises.

timvandermeij committed:
Fix code duplication in the rasterization logic in `test/driver.js`

Now that the rasterization logic is encapsulated in a class, we can
easily move the container creation into a separate static method.

Snuffleupagus commented on test/driver.js:
```diff
@@ -143,106 +145,75 @@ async function resolveImages(node, silentErrors = false) {
+        promise: null,
+      },
+    };
+    return shadow(Rasterize, "annotationStylePromise", loadStyles(styles));
```
Given that you've added proper shadowing, do we still need to track the promises manually here?
Basically the pre-existing code feels overly complicated, and I wonder if we couldn't now do something like
```javascript
const styles = {
  common: "../web/annotation_layer_builder.css",
  overrides: "./annotation_layer_builder_overrides.css",
}
return shadow(this, "annotationStylePromise", loadStyles(styles));
```
with `loadStyles` modified along e.g. these lines:
```javascript
function loadStyles(styles) {
  const promises = [];

  for (const type of styles) {
    const file = styles[type];

    promises.push(new Promise(function (resolve, reject) {
      const xhr = new XMLHttpRequest();
      xhr.open("GET", file);
      xhr.onload = function () {
        resolve(xhr.responseText);
      };
      xhr.onerror = function (e) {
        reject(new Error(`Error fetching style (${file}): ${e}`));
      };
      xhr.send(null);
    });
  }

  return Promise.all(promises);
}
```

timvandermeij replied:
Good idea; I have made the changes. The driver file is now exactly 100 lines shorter than before :-)

Snuffleupagus commented on test/driver.js:
```diff
@@ -260,13 +200,12 @@ const rasterizeAnnotationLayer = (function rasterizeAnnotationLayerClosure() {
 
       // Rendering annotation layer as HTML.
-      const [common, overrides] = await stylePromise;
+      const [common, overrides] = await Rasterize.annotationStylePromise;
```
Please note that `static` methods can *directly* call other `static` methods and access static properties.
Hence the more compact `this.annotationStylePromise` format could be used here, and similarly elsewhere in file.

Snuffleupagus reviewed (approved):
r=me, with one final comment addressed (everywhere that it applies); thanks a lot for doing this!

I'd suggest running the tests maybe a couple of times, to ensure that there's no new intermittent failures (since the usage of `async` perhaps affect timings ever so slightly).

timvandermeij commented:
/botio test

Snuffleupagus commented:
/botio browsertest

Snuffleupagus commented:
/botio browsertest
